File: packages/silo-client-java/src/test/java/in/org/quicko/silo/client/TransportTest.java
package in.org.quicko.silo.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import in.org.quicko.silo.client.errors.ConflictException;
import in.org.quicko.silo.client.errors.ForbiddenException;
import in.org.quicko.silo.client.errors.InvalidResponseException;
import in.org.quicko.silo.client.errors.MediaInUseException;
import in.org.quicko.silo.client.errors.NetworkException;
import in.org.quicko.silo.client.errors.NotFoundException;
import in.org.quicko.silo.client.errors.RequestAbortedException;
import in.org.quicko.silo.client.errors.RequestTimeoutException;
import in.org.quicko.silo.client.errors.SiloException;
import in.org.quicko.silo.client.errors.UnauthorizedException;
import in.org.quicko.silo.client.errors.ValidationFailedException;
import in.org.quicko.silo.client.instance.HealthReport;
import in.org.quicko.silo.client.support.RecordedRequest;
import in.org.quicko.silo.client.support.StubHttp;
import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.time.Duration;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class TransportTest {
  private static final String Health = "{\"status\":\"ok\",\"version\":\"1.1.0\"}";

  private StubHttp server;

  @BeforeEach
  void newStub() {
    server = StubHttp.create();
  }

  @Test
  void readsTheOneRouteThatIsNeverAuthenticated() {
    server.enqueueJson(Health);
    HealthReport report = server.anonymous().health();

    assertEquals("ok", report.status());
    assertEquals("1.1.0", report.version());
    assertEquals("/api/health", server.lastRequest().path());
  }

  @Test
  void sendsTheKeyAsABearerToken() {
    server.enqueueJson(Health);
    server.silo("silo_key_abc").health();

    assertEquals("Bearer silo_key_abc", server.lastRequest().header("authorization"));
  }

  @Test
  void sendsNoAuthorizationHeaderWithoutAKeyBecauseAnonymousReadsAreLegal() {
    server.enqueueJson(Health);
    server.anonymous().health();

    assertNull(server.lastRequest().header("authorization"));
  }

  @Test
  void stripsTrailingSlashesFromTheBaseUrlSoThePathIsNotDoubled() {
    server.enqueueJson(Health);
    new Silo(SiloOptions.of(StubHttp.BaseUrl + "///").httpClient(server.client())).health();

    assertEquals("/api/health", server.lastRequest().path());
  }

  @Test
  void sendsCallerHeadersAndKeepsAuthorizationItsOwn() {
    server.enqueueJson(Health);
    new Silo(SiloOptions.of(StubHttp.BaseUrl, "silo_key_abc")
        .headers(Map.of("X-Trace", "abc123", "Authorization", "Bearer wrong"))
        .httpClient(server.client()))
        .health();

    RecordedRequest request = server.lastRequest();
    assertEquals("abc123", request.header("x-trace"));
    assertEquals("Bearer silo_key_abc", request.header("authorization"));
  }

  @Test
  void raisesTheClassTheWireCodeNames() {
    assertInstanceOf(ValidationFailedException.class, refusal(400, "validation_failed"));
    assertInstanceOf(UnauthorizedException.class, refusal(401, "unauthorized"));
    assertInstanceOf(ForbiddenException.class, refusal(403, "forbidden"));
    assertInstanceOf(NotFoundException.class, refusal(404, "not_found"));
    assertInstanceOf(ConflictException.class, refusal(409, "conflict"));
  }

  @Test
  void carriesTheValidationDetailTheServerNamed() {
    server.enqueue(400, "application/json",
        "{\"error\":{\"code\":\"validation_failed\",\"message\":\"bad\",\"details\":"
            + "[{\"path\":\"/title\",\"message\":\"required\"}]}}");

    ValidationFailedException caught =
        assertThrows(ValidationFailedException.class, () -> server.silo("k").health());

    assertEquals(1, caught.details().size());
    assertEquals("/title", caught.details().get(0).path());
    assertEquals("required", caught.details().get(0).message());
  }

  @Test
  void readsTheObjectShapedDetailsOfAMediaInUseRefusal() {
    server.enqueue(409, "application/json",
        "{\"error\":{\"code\":\"media_in_use\",\"message\":\"still referenced\",\"details\":"
            + "{\"usage_count\":3,\"visible_count\":2,\"visible_capped\":true,\"referrers\":"
            + "[{\"media_id\":\"m1\",\"project\":\"acme\",\"env\":\"prod\","
            + "\"collection\":\"posts\",\"entry_id\":\"e1\"}]}}}");

    MediaInUseException caught =
        assertThrows(MediaInUseException.class, () -> server.silo("k").health());

    assertEquals(3, caught.usageCount());
    assertEquals(2, caught.visibleCount());
    assertTrue(caught.visibleCapped());
    assertEquals("prod", caught.referrers().get(0).environment());
    assertInstanceOf(ConflictException.class, caught);
  }

  @Test
  void fallsBackOnTheStatusWhenTheBodyIsNotJsonAtAll() {
    server.enqueue(403, "text/html", "<html>proxy denied</html>");

    ForbiddenException caught =
        assertThrows(ForbiddenException.class, () -> server.silo("k").health());
    assertEquals("<html>proxy denied</html>", caught.getMessage());
  }

  @Test
  void keepsAnUnknownCodeAsASiloExceptionRatherThanGuessing() {
    server.enqueue(418, "application/json",
        "{\"error\":{\"code\":\"teapot\",\"message\":\"short and stout\"}}");

    SiloException caught = assertThrows(SiloException.class, () -> server.silo("k").health());
    assertEquals(418, caught.status());
    assertEquals("unknown", caught.code());
  }

  @Test
  void refusesAJsonRouteThatAnsweredSomethingElse() {
    server.enqueue(200, "text/html", "<html>a proxy page</html>");
    assertThrows(InvalidResponseException.class, () -> server.silo("k").health());
  }

  @Test
  void tellsADeadlineFromAHostThatWasNeverThere() {
    server.enqueueFailure(new SocketTimeoutException("read timed out"));
    RequestTimeoutException timedOut = assertThrows(RequestTimeoutException.class,
        () -> server.silo("k").health(RequestOptions.within(Duration.ofMillis(200))));
    assertEquals(Duration.ofMillis(200), timedOut.timeout());

    server.enqueueFailure(new ConnectException("connection refused"));
    NetworkException unreachable =
        assertThrows(NetworkException.class, () -> server.silo("k").health());
    assertTrue(unreachable.getMessage().contains("connection refused"),
        "what the call failed with belongs in the message, not only in the cause");
  }

  @Test
  void tellsACallersOwnCancellationFromADeadline() {
    CancellationSignal signal = CancellationSignal.create();
    signal.cancel();

    assertThrows(RequestAbortedException.class,
        () -> server.silo("k").health(RequestOptions.until(signal)));
    assertEquals(0, server.requestCount(), "a cancelled call is not sent at all");
  }

  @Test
  void doesNotRetryAnythingItself() {
    server.enqueueFailure(new ConnectException("connection refused"));
    assertThrows(NetworkException.class, () -> server.silo("k").health());

    assertEquals(1, server.requestCount(),
        "a retried POST is a duplicate entry, so retry policy is the caller's");
  }

  private Throwable refusal(int status, String code) {
    server.enqueue(status, "application/json",
        "{\"error\":{\"code\":\"" + code + "\",\"message\":\"refused\"}}");
    return assertThrows(SiloException.class, () -> server.silo("k").health());
  }
}

File: packages/silo-client-dart/test/transport_test.dart
import 'dart:async';

import 'package:http/http.dart' as http;
import 'package:silo_client/silo_client.dart';
import 'package:test/test.dart';

import 'support/stub_http.dart';

void main() {
  late StubHttp stub;
  setUp(() => stub = StubHttp());

  test('sends the key as a bearer token, and nothing when anonymous', () async {
    stub
      ..json({'status': 'ok', 'version': '1.4.0'})
      ..json({'status': 'ok', 'version': '1.4.0'});

    final health = await stub.silo().health();
    await stub.silo(key: null).health();

    expect(health.version, '1.4.0');
    expect(stub.requests[0].url.toString(), 'http://silo.test/api/health');
    expect(stub.requests[0].headers['authorization'], 'Bearer test-key');
    expect(stub.requests[1].headers.containsKey('authorization'), isFalse);
  });

  test('encodes a path segment the way encodeURIComponent does', () async {
    stub.json({'items': <Object?>[]});
    await stub.silo().project('a b/c').environments.list();
    expect(stub.last.url.path, '/api/projects/a%20b%2Fc/envs');
  });

  test('sends a JSON body with its content type', () async {
    stub.json({'id': '01P', 'name': 'acme'});
    final project = await stub.silo().projects.create('acme');

    expect(project.id, '01P');
    expect(stub.last.headers['content-type'], startsWith('application/json'));
    expect(stub.lastBody, {'id': 'acme'});
  });

  test('maps each wire code to its exception, and falls back on the status', () async {
    final silo = stub.silo();
    stub
      ..json({
        'error': {
          'code': 'validation_failed',
          'message': 'bad',
          'details': [
            {'path': '/title', 'message': 'required'},
          ],
        },
      }, status: 400)
      ..json({
        'error': {'code': 'not_found', 'message': 'gone'},
      }, status: 404)
      ..json({
        'error': {'code': 'something_new', 'message': 'stale'},
      }, status: 409)
      ..text('<html>bad gateway</html>', status: 502);

    await expectLater(
      silo.projects.list(),
      throwsA(isA<ValidationFailedException>().having((caught) => caught.details.single.path, 'path', '/title')),
    );
    await expectLater(silo.projects.list(), throwsA(isA<NotFoundException>()));
    await expectLater(silo.projects.list(), throwsA(isA<ConflictException>()));
    await expectLater(
      silo.projects.list(),
      throwsA(isA<SiloException>().having((caught) => caught.status, 'status', 502)),
    );
  });

  test('refuses a 2xx that is not the JSON the route promises', () async {
    stub.text('<html></html>');
    await expectLater(stub.silo().projects.list(), throwsA(isA<InvalidResponseException>()));
  });

  test('a deadline raises RequestTimeoutException', () async {
    stub.hang();
    await expectLater(
      stub.silo().projects.list(const RequestOptions(timeout: Duration(milliseconds: 10))),
      throwsA(isA<RequestTimeoutException>()),
    );
  });

  test('the client default deadline applies when a call sets none', () async {
    stub.hang();
    await expectLater(
      stub.silo(timeout: const Duration(milliseconds: 10)).projects.list(),
      throwsA(isA<RequestTimeoutException>()),
    );
  });

  test('a cancellation raises RequestAbortedException, before or during the call', () async {
    final cancelled = CancellationSignal()..cancel();
    await expectLater(
      stub.silo().projects.list(RequestOptions(cancellation: cancelled)),
      throwsA(isA<RequestAbortedException>()),
    );
    expect(stub.requests, isEmpty);

    stub.hang();
    final signal = CancellationSignal();
    final call = stub.silo().projects.list(RequestOptions(cancellation: signal));
    Timer.run(signal.cancel);
    await expectLater(call, throwsA(isA<RequestAbortedException>()));
  });

  test('a request that never lands raises NetworkException naming the cause', () async {
    stub.fail(http.ClientException('connection refused'));
    await expectLater(
      stub.silo().projects.list(),
      throwsA(isA<NetworkException>().having((caught) => '$caught', 'message', contains('connection refused'))),
    );
  });

  test('withKey keeps the server and swaps the key', () async {
    stub.json({'items': <Object?>[]});
    await stub.silo().withKey('other').projects.list();
    expect(stub.last.headers['authorization'], 'Bearer other');
    expect(stub.last.url.host, 'silo.test');
  });

  test('refuses an empty URL', () {
    expect(() => Silo(const SiloOptions(url: ' ')), throwsArgumentError);
  });
}
